### PR TITLE
arch/risc-v/bl602 : add gpioirq、 i2c(master) driver.

### PR DESCRIPTION
**Pull request by liangzhanggb**
## Summary
arch/risc-v/bl602 : add gpioirq、 i2c(master) driver.

## Impact
N/A

## Testing
config:
boards/risc-v/bl602/bl602evb/configs/gpio/defconfig
boards/risc-v/bl602/bl602evb/configs/i2c/defconfig

test gpio:
gpio /dev/gpin0
gpio -o 0 /dev/gpout1
gpio -o 1 /dev/gpout1
gpio -w 1 /dev/gpint2

test i2c:(support master, not support slave)
i2c set -a50 -r09 -f500 33
i2c get -a50 -r09

## References:
https://github.com/apache/incubator-nuttx/pull/2637  @btashton

### CODE REVIEW

**Comment by btashton** [comment]
@liangzhanggb I will look but I had already created a WIP PR for i2c and last night was testing the rest of the full implementation. It would be nice to not be duplicating work.

File: boards/risc-v/bl602/bl602evb/src/bl602_bringup.c [hunk 1]
```diff
@@ -149,6 +151,23 @@ int bl602_bringup(void)
...
+  ret = bl602_gpio_initialize();
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "Failed to initialize GPIO Driver: %d\n", ret);
+      return ret;
+    }
+#endif
+
+#ifdef CONFIG_I2C 
+  ret = bl602_i2cbus_initialize(0);
```
Comment (btashton) [inline boards/risc-v/bl602/bl602evb/src/bl602_bringup.c:164]: Can you also hook in the i2ctool support here (you can see it in my PR).  Makes it easier to do testing.

File: boards/risc-v/bl602/bl602evb/src/bl602_gpio.c [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+  uint32_t tmp_val;
+
+  DEBUGASSERT(bl602xgpint != NULL && value != NULL);
+  DEBUGASSERT(bl602xgpint->bl602xgpio.id < BOARD_NGPIOINT);
+  gpioinfo("Reading int pin...\n");
+
+  gpio_pin = (g_gpiointinputs[bl602xgpint->bl602gpio.id] & GPIO_PIN_MASK) >>
+    GPIO_PIN_SHIFT;
+  tmp_val = getreg32(BL602_GPIO_CFGCTL30 + ((gpio_pin >> 5) << 2));
+  pos = gpio_pin % 32;
```
Comment (btashton) [inline boards/risc-v/bl602/bl602evb/src/bl602_gpio.c:503]: Why the % 32?

File: boards/risc-v/bl602/bl602evb/src/bl602_gpio.c [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+  gpioinfo("Writing %d\n", (int)value);
+
+  uint8_t gpio_pin = (g_gpiooutputs[bl602xgpio->id] & GPIO_PIN_MASK) >>
+    GPIO_PIN_SHIFT;
+  uint32_t *p_out    = (uint32_t *)(BL602_GPIO_CFGCTL32 +
+                        ((gpio_pin >> 5) << 2));
+  uint32_t  pos      = gpio_pin % 32;
+  uint32_t  tmp_out;
+
+  tmp_out = *p_out;
```
Comment (btashton) [inline boards/risc-v/bl602/bl602evb/src/bl602_gpio.c:465]: This block should be locked to protect the modify operation.

File: boards/risc-v/bl602/bl602evb/src/bl602_gpio.c [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+
+  DEBUGASSERT(bl602xgpio != NULL);
+  DEBUGASSERT(bl602xgpio->id < BOARD_NGPIOOUT);
+  gpioinfo("Writing %d\n", (int)value);
+
+  uint8_t gpio_pin = (g_gpiooutputs[bl602xgpio->id] & GPIO_PIN_MASK) >>
+    GPIO_PIN_SHIFT;
+  uint32_t *p_out    = (uint32_t *)(BL602_GPIO_CFGCTL32 +
+                        ((gpio_pin >> 5) << 2));
+  uint32_t  pos      = gpio_pin % 32;
```
Comment (btashton) [inline boards/risc-v/bl602/bl602evb/src/bl602_gpio.c:462]: Why is this modulo 32?

File: boards/risc-v/bl602/bl602evb/src/bl602_gpio.c (hunk 3 above)
Comment (btashton) [reply boards/risc-v/bl602/bl602evb/src/bl602_gpio.c:465]: Looks like you could just drop tmp_out and pos and just call modify on p_out

File: arch/risc-v/src/bl602/bl602_i2c.c [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+  uint8_t bclk_div = 0;
+
+  bclk_div = bl602_glb_get_bclk_div();
+  if (clk >= 100000)
+    {
+      bl602_set_i2c_clk(1, 0);
+      bl602_i2c_setprd((SYS_TEMCORE_CLOCK_GET / (bclk_div + 1)) / (clk * 4) -
+                       1);
+    }
+  else if (clk >= 8000)
```
Comment (btashton) [inline arch/risc-v/src/bl602/bl602_i2c.c:633]: Why are we supporting these lower frequencies?  I am not aware of an i2c device that operates so slow.  Usually you just have 100k 400k and some special devices that can go up to 5MHz

File: arch/risc-v/src/bl602/bl602_i2c.c [hunk 6]
```diff
@@ -0,0 +1,1049 @@
...
+};
+
+/* I2C device structures */
+
+#ifdef CONFIG_BL602_I2C0
+static const struct bl602_i2c_config_s bl602_i2c0_config =
+{
+  .reg_base = BL602_I2C_BASE,
+  .irq      = BL602_IRQ_I2C,
+  .clk_freq = 500,
```
Comment (btashton) [inline arch/risc-v/src/bl602/bl602_i2c.c:146]: Why 500Hz?

**Review by btashton — CHANGES_REQUESTED** [review]
Some comments here, I commented on one, but there are a few printf in the i2c driver that need to be change to the debug versions.  I have a display and a sensor wired up that I was using for my testing I'll test this after work on that.

**Comment by liangzhanggb** [comment]
> @liangzhanggb I will look but I had already created a WIP PR for i2c and last night was testing the rest of the full implementation. It would be nice to not be duplicating work.

Great, thank you very much, I am going to close my i2c_drv PR, and then create a gpio_drv PR

**Comment by btashton** [comment]
@liangzhanggb let's go with yours it's more complete. Maybe we can coordinate more on other drivers.

**Comment by yamt** [comment]
please avoid using non-ascii characters in the commit message. ("、")

File: arch/risc-v/src/bl602/bl602_i2c.c (hunk 6 above)
Comment (liangzhanggb) [reply arch/risc-v/src/bl602/bl602_i2c.c]: I have fixed it ,Thanks.

arch/risc-v/src/bl602/bl602_i2c.c
#define I2C_DEFAULT_FREQUENCY 100000

File: boards/risc-v/bl602/bl602evb/src/bl602_gpio.c (hunk 4 above)
Comment (liangzhanggb) [reply boards/risc-v/bl602/bl602evb/src/bl602_gpio.c]: I have fixed it ,Thanks.

File: arch/risc-v/src/bl602/bl602_i2c.c (hunk 5 above)
Comment (liangzhanggb) [reply arch/risc-v/src/bl602/bl602_i2c.c:634]: BL602 I2c clk Ranges: 305Hz~3MHz

100~400K, test pass( use device)

3MHz, test pass (use oscilloscope )

**Comment by btashton** [comment]
@liangzhanggb Let's merge this in, we can make some more improvements later.

**Pull request merged**